README.md
# torchlambertw: Lambert W function and Lambert W x F distributions in pytorch


![Python](https://img.shields.io/badge/python-3670A0?style=for-the-badge&logo=python&logoColor=ffdd54)
![PyTorch](https://img.shields.io/badge/PyTorch-%23EE4C2C.svg?style=for-the-badge&logo=PyTorch&logoColor=white)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](http://makeapullrequest.com)
[![MIT license](https://img.shields.io/badge/License-MIT-blue.svg)](https://lbesson.mit-license.org/)
![Github All Releases](https://img.shields.io/github/downloads/gmgeorg/pypsps/total.svg)


**IMPORTANT**: This is *NOT* ready for use.  This is just a prototype serving as basis of discussion in https://github.com/pytorch/pytorch/issues/108948. 

This has not been properly tested w/ all pytorch functionality and should not be used other than for prototyping/R&D.

See https://github.com/gmgeorg/torchlambertw/issues for remaining issues/TODOs.

---

# Overview

This library is a native implementation in `pytorch` of

 * the Lambert W function (`special.lambertw`)

 * Lambert W x F distributions (`torch.distributions`)

While this library is for now standalone, the goal is to get both the mathematicla function as well as the distributions into `torch` core package.

See also https://github.com/pytorch/pytorch/issues/108948.


## Installation

It can be installed directly from GitHub using:

```python
pip install git+https://github.com/gmgeorg/torchlambertw.git
```


## Lambert W function (math)

Implementation of the Lambert W function (special function) in `torch`:

```python
import torchlambertw as tw
import numpy as np
special.lambertw(torch.tensor([-1., 0., 1., -np.exp(-1)]))
```
output:
```bash
tensor([nan,  0.0000,  0.5671, -1.0000], dtype=torch.float64)
```

As a more interesting example you can use this implementation to replicate the figure on the [Lambert W Function](https://en.wikipedia.org/wiki/Lambert_W_function) Wikipedia page:

```python
import numpy as np
import matplotlib.pyplot as plt
from torchlambertw import special

def plot_lambertW(range_start, range_end, num_points=2000):
    x_values = np.linspace(range_start, range_end, num_points)
    x_values_torch = torch.tensor(x_values)
    principal_branch_values = special.lambertw(x_values_torch, k=0).numpy()
    non_principal_branch_values = special.lambertw(x_values_torch, k=-1).numpy()
    
    plt.figure(figsize=(8, 5))
    plt.plot(x_values, principal_branch_values, label="Principal Branch", color='blue')
    plt.plot(x_values, non_principal_branch_values, label="Non-Principal Branch", color='red')
    
    plt.title("Lambert W Function")
    plt.xlabel("x")
    plt.ylabel("W(x)")
    plt.xlim(range_start, range_end)
    plt.ylim(-4, 2)  # same range as wiki figure
    plt.axhline(0, color='black', linestyle='--', linewidth=0.5)
    plt.axvline(0, color='black', linestyle='--', linewidth=0.5)
    plt.legend()
    
    plt.grid(True)
    plt.show()

# Example usage:
plot_lambertW(-1, 6)

```
![Lambert W Function](imgs/lambertw_plot.png)

## Lambert W x F distributions

For the original papers see Goerg 2011 & 2015. If you want to jump into applications and examples I suggest looking at the [**LambertW** R package](https://github.com/gmgeorg/LambertW) for detailed references and links to many external examples on Stackoverflow / cross-validated and other external blogs.


### Heavy-tail distributions first
```python
from torchlambertw import distributions as tlwd

# Implements a Lambert W x Normal distribution with (loc=1, scale=3, tailweight=0.75)
m = tlwd.LambertWNormal(loc=torch.tensor([1.0]), 
                        scale=torch.tensor([3.0]), 
                        tailweight=torch.tensor([0.75]))
m.sample((2,)) 
```
```
tensor([[ 0.0159], [-0.9322]])
```

This distribution is quite heavy-tailed with moments existing only up to `1 / tailweight = 1.33`, ie this random variable / distribution has infinite (population) variance.

```python
m.tailweight, m.support, m.mean, m.variance
```
```
(tensor([0.7500]), Real(), tensor([1.]), tensor([inf]))
```

Let's draw a random sample from distribution and plot density / ecdfplot.

```python
torch.manual_seed(0)
# Use a less heavy-tailed distribution with a tail parameter of 0.25 (ie moments < 1/0.25 = 4 exist).
m = tlwd.LambertWNormal(loc=torch.tensor([1.0]), scale=torch.tensor([3.0]), 
                        tailweight=torch.tensor([0.25]))
y = m.sample((1000,)).numpy().ravel()

import seaborn as sns
import statsmodels.api as sm

sns.displot(y, kde=True)
plt.show()
sm.qqplot(y, line='45', fit=True)
plt.grid()
plt.show()
```
![Lambert W x Gaussian histogram and KDE](imgs/lambertw_gauss_hist_kde.png)

![Lambert W x Gaussian qqnorm plot](imgs/lambertw_gauss_qqnorm.png)

#### Back-transformation

The parameters `(loc, scale, tailweight)` can be estimated from the data (see **LambertW** R package or also the `gaussianize` package for an **sklearn** wrapper) -- see also https://github.com/gmgeorg/torchlambertw/issues/2.

Let's assume you have the parameters estimated; then you can use this to obtain the unobserved, Gaussian data:

```python
torch.manual_seed(0)

m = tlwd.LambertWNormal(loc=torch.tensor([1.0]), scale=torch.tensor([3.0]), 
                        tailweight=torch.tensor([0.25]))

y = m.sample((1000,)).numpy().ravel()
x = m.transforms[0]._inverse(torch.tensor(y)).numpy().ravel()
sns.displot(x, kde=True)
plt.show()
sm.qqplot(x, line='45', fit=True)
plt.grid()
plt.show()

```

![Lambert W x Gaussian histogram and KDE](imgs/lambertw_gauss_latent_hist_kde.png)

![Lambert W x Gaussian qqnorm plot](imgs/lambertw_gauss_latent_qqplot.png)
### Skewed Lambert W x F distributions

For introducting skewness one can use the skewed Lambert W x F distributions; particularly
interesting are skewed Lambert W x Gaussian distributions as they can be used to transform Normal data to skewed data (and back).

```python
torch.manual_seed(0)
m = tlwd.SkewLambertWNormal(loc=torch.tensor([1.0]), scale=torch.tensor([3.0]), 
                        skewweight=torch.tensor([0.25]))
y = m.sample((1000,)).numpy().ravel()


sns.displot(y, kde=True)
plt.show()
sm.qqplot(y, line='45', fit=True)
plt.grid()
plt.show()
```

![skewed Lambert W x Gaussian histogram and KDE](imgs/lambertw_skew_gauss_hist_kde.png)

![skewed Lambert W x Gaussian qqnorm plot](imgs/lambertw_skew_gauss_qqplot.png)

#### Skewed Lambert W x Exponential distributions

Lambert W x F distribution can be generated for any arbitratry F.  Here we show an example of a skewed Lambert W x Exponential distribution. For `skewweight > 0` , this will be more (right) skewed than a baseline exponential distribution. 

See Goerg (2011) and Kaarik et al. (2023).
```python
torch.manual_seed(0)
m = tlwd.SkewLambertWExponential(rate=torch.tensor([3.0]), 
                                 skewweight=torch.tensor([0.2]))
y = m.sample((1000,)).numpy().ravel()
print("mean: theoretical %f; empirical: %f" % (m.mean.numpy()[0], np.mean(x)))

sns.displot(y, kde=True)
plt.show()
sm.qqplot(y, line='45', fit=True, dist=stats.expon)
plt.grid()
plt.show()

```

Mean is larger than the 1/3. rate of the baseline exponential distribution because of the longer tail.
```
mean: theoretical 0.520833; empirical: 0.481116
```

![skewed Lambert W x Exponential histogram and KDE](imgs/lambertw_skew_exp_hist_kde.png)
![skewed Lambert W x Exponential qqnorm plot](imgs/lambertw_skew_exp_qqplot.png)


## Implementation details

This implementation closely follows the TensorFlow Probability version in [`tfp.special.lambertw`](https://www.tensorflow.org/probability/api_docs/python/tfp/math/lambertw).


See also [here](https://github.com/thibsej/unbalanced-ot-functionals/blob/13f2203b3993d973f929578085ea458c5c1a7a78/common/torch_lambertw.py) and [here](
https://github.com/AminJun/BreakingCertifiableDefenses/blob/cc469fa48f7efba21f3584e233c4db0c9a4856c1/RandomizedSmoothing/projected_sinkhorn/lambertw.py
)) for minimum example `pytorch` implementations.


## References

* Corless, R.M., et al. (1996) On the LambertW Function. Advances in Computational Mathematics, 5, 329-359.
https://doi.org/10.1007/BF02124750 

* Lambert W implementation in TensorFlow: https://www.tensorflow.org/probability/api_docs/python/tfp/math/lambertw

* Goerg (2011). *Lambert W random variables—a new family of generalized skewed distributions with applications to risk estimation.* Ann. Appl. Stat. 5 (3) 2197 - 2230, 2011. https://doi.org/10.1214/11-AOAS457 

* Goerg (2015) *The Lambert Way to Gaussianize Heavy-Tailed Data with the Inverse of Tukey’s h Transformation as a Special Case*. The Scientific World Journal. Volume 2015 | Article ID 909231 | https://doi.org/10.1155/2015/909231

* Käärik, Meelis & Selart, Anne & Puhkim, Tuuli & Tee, Liivika. (2023). *Lambert W random variables and their applications in loss modelling*. https://arxiv.org/pdf/2307.05644.pdf

# License

This project is licensed under the terms of the [MIT license](LICENSE).


CHANGELOG.md
# CHANGELOG.md

See the issues page for todos and upcoming changes.

## 0.0.1-alpha (unreleased)

This is the very first attempt at providing a native torch implementation of the Lambert W function and Lambert W x distriubtions.

See `README` for reference on original works and the `pytorch` issue referencing the missing functionality.


torchlambertw/distributions.py
"""Module for Lambert W x F distributions."""

import torch

from . import special
from . import transforms
import torch.distributions as td


class LambertWDistribution(td.transformed_distribution.TransformedDistribution):
    r"""
    Creates a heavy-tail Lambert W x Normal distribution parameterized by
    `shift`, `scale`, and `tailweight` where::

        X ~ F(shift, scale)
        U = (X - shift) / scale
        Z = U * exp(tailweight / 2. * U^2)
        Y = Z * scale + shift
        Y ~ Lambert W x F(shift, scale, tailweight)

    Args:
        shift (float or Tensor): location/shift parameter of the latent F distribution.
          For (non-negative) scale-family distributions (e.g., exponential or gamma)
          this should be set to 0.
        scale (float or Tensor): scale parameter of latent F distribution
        tailweight (float or Tensor): tailweight ("delta") of the Lambert W x F distribution.
          If 0., then it reduces to a F(loc, scale) distribution.
    """
    arg_constraints = {
        "shift": td.constraints.real,
        "scale": td.constraints.positive,
        "tailweight": td.constraints.greater_than_eq(0.0),
    }
    support = td.constraints.real
    has_rsample = True

    def __init__(
        self,
        base_distribution,
        base_dist_args,
        shift,
        scale,
        tailweight,
        validate_args=None,
    ):
        self.tailweight = tailweight
        self.shift = shift
        super().__init__(
            base_distribution=base_distribution(
                **base_dist_args, validate_args=validate_args
            ),
            transforms=transforms.LambertWTailTransform(
                shift=shift, scale=scale, tailweight=tailweight
            ),
            validate_args=validate_args,
        )

    def expand(self, batch_shape, _instance=None):
        new = self._get_checked_instance(LambertWDistribution, _instance)
        return super().expand(batch_shape, _instance=new)

    @property
    def loc(self):
        return self.transforms[0].shift

    @property
    def scale(self):
        return self.transforms[0].scale

    @property
    def mean(self):
        pass

    @property
    def mode(self):
        pass

    @property
    def variance(self):
        pass


class LambertWNormal(td.transformed_distribution.TransformedDistribution):
    r"""
    Creates a heavy-tail Lambert W x Normal distribution parameterized by
    `loc`, `scale`, and `tailweight` where::

        X ~ Normal(loc, scale)
        U = (X - loc) / scale
        Z = U * exp(tailweight / 2. * U^2)
        Y = Z * scale + loc
        Y ~ Lambert W x Normal(loc, scale, tailweight)

    Args:
        loc (float or Tensor): mean of latent Normal distribution
        scale (float or Tensor): standard deviation of latent Normal distribution
        tailweight (float or Tensor): tailweight ("delta") of the Lambert W x Normal distribution.
          If 0., then it reduces to a Normal(loc, scale) distribution.
    """
    arg_constraints = {
        "loc": td.constraints.real,
        "scale": td.constraints.positive,
        "tailweight": td.constraints.greater_than_eq(0.0),
    }
    support = td.constraints.real
    has_rsample = True

    def __init__(self, loc, scale, tailweight, validate_args=None):
        self.tailweight = tailweight
        super().__init__(
            base_distribution=td.Normal(
                loc=loc, scale=scale, validate_args=validate_args
            ),
            transforms=transforms.LambertWTailTransform(
                shift=loc, scale=scale, tailweight=tailweight
            ),
            validate_args=validate_args,
        )

    def expand(self, batch_shape, _instance=None):
        new = self._get_checked_instance(LambertWNormal, _instance)
        return super().expand(batch_shape, _instance=new)

    @property
    def loc(self):
        return self.base_dist.loc

    @property
    def scale(self):
        return self.base_dist.scale

    @property
    def mean(self):
        loc_abs_times_tailweight = (self.base_dist.loc.abs() + 1) * (
            1.0 - self.tailweight
        )
        return torch.where(loc_abs_times_tailweight > 0, self.base_dist.loc, torch.nan)

    @property
    def mode(self):
        return self.base_dist.loc

    @property
    def variance(self):
        z_variance = 1.0 / torch.pow(1 - 2 * self.tailweight, 1.5)
        y_variance = z_variance * self.base_dist.scale.pow(2)
        return torch.where(y_variance > 0, y_variance, torch.inf)


class SkewLambertWDistribution(td.transformed_distribution.TransformedDistribution):
    r"""
    Creates a skewed Lambert W x Normal distribution parameterized by
    `shift`, `scale`, and `skewweight` where::

        X ~ F(shift, scale)
        U = (X - shift) / scale
        Z = U * exp(skewweight * U)
        Y = Z * scale + shift
        Y ~ Lambert W x F(shift, scale, tailweight)

    Args:
        shift (float or Tensor): location/shift parameter of the latent F distribution.
          For (non-negative) scale-family distributions (e.g., exponential or gamma)
          this should be set to 0.
        scale (float or Tensor): scale parameter of latent F distribution
        tailweight (float or Tensor): tailweight ("delta") of the Lambert W x F distribution.
          If 0., then it reduces to a F(shift, scale) distribution.
    """
    arg_constraints = {
        "shift": td.constraints.real,
        "scale": td.constraints.positive,
        "skewweight": td.constraints.real,
    }
    support = td.constraints.real
    has_rsample = True

    def __init__(
        self,
        base_distribution,
        base_dist_args,
        shift,
        scale,
        skewweight,
        validate_args=None,
    ):
        self.skewweight = skewweight
        self.shift = shift
        super().__init__(
            base_distribution=base_distribution(
                **base_dist_args, validate_args=validate_args
            ),
            transforms=transforms.LambertWSkewTransform(
                shift=shift, scale=scale, skewweight=skewweight
            ),
            validate_args=validate_args,
        )

    def expand(self, batch_shape, _instance=None):
        new = self._get_checked_instance(LambertWDistribution, _instance)
        return super().expand(batch_shape, _instance=new)

    @property
    def loc(self):
        return self.transforms[0].shift

    @property
    def scale(self):
        return self.transforms[0].scale

    @property
    def mean(self):
        pass

    @property
    def mode(self):
        pass

    @property
    def variance(self):
        pass


class SkewLambertWNormal(td.transformed_distribution.TransformedDistribution):
    r"""
    Creates a skewed Lambert W x Normal distribution parameterized by
    `loc`, `scale`, and `skewweight` where::

        X ~ Normal(loc, scale)
        U = (X - loc) / scale
        Z = U * exp(skewweight * U)
        Y = Z * scale + loc
        Y ~ Lambert W x Normal(loc, scale, skewweight)

    Args:
        loc (float or Tensor): mean of latent Normal distribution
        scale (float or Tensor): standard deviation of latent Normal distribution
        skewweight (float or Tensor): skewweight ("gamma") of the Lambert W x Normal distribution.
          If 0., then it reduces to a Normal(loc, scale) distribution.
    """
    arg_constraints = {
        "loc": td.constraints.real,
        "scale": td.constraints.positive,
        "skewweight": td.constraints.real,
    }
    support = td.constraints.real
    has_rsample = True

    def __init__(self, loc, scale, skewweight, validate_args=None):
        self.skewweight = skewweight
        super().__init__(
            base_distribution=td.Normal(
                loc=loc, scale=scale, validate_args=validate_args
            ),
            transforms=transforms.LambertWSkewTransform(
                shift=loc, scale=scale, skewweight=skewweight
            ),
            validate_args=validate_args,
        )

    def expand(self, batch_shape, _instance=None):
        new = self._get_checked_instance(LambertWNormal, _instance)
        return super().expand(batch_shape, _instance=new)

    @property
    def loc(self):
        return self.base_dist.loc

    @property
    def scale(self):
        return self.base_dist.scale

    @property
    def mean(self):
        # See eq (4.1) in Goerg 2011.
        mu_y = self.base_dist.loc + self.base_dist.scale * self.skewweight * torch.exp(
            0.5 * self.skewweight ** 2
        )
        return mu_y

    @property
    def mode(self):
        pass

    @property
    def variance(self):
        # See eq (4.2) in Goerg 2011.
        gamma_sq = self.skewweight ** 2
        exp_gamma_sq = torch.exp(gamma_sq)
        sigma2_y = (
            self.base_dist.scale.pow(2)
            * exp_gamma_sq
            * ((4 * gamma_sq + 1) * exp_gamma_sq - gamma_sq)
        )
        return sigma2_y


class SkewLambertWExponential(td.transformed_distribution.TransformedDistribution):
    r"""
    Creates a skewed Lambert W x Exponential distribution parameterized by
    `rate` (= 1 / scale), and `skewweight` where::

        X ~ Exponential(rate)
        U = X * rate ( = X / scale)
        Z = U * exp(skewweight * U)
        Y = Z / rate
        Y ~ Lambert W x Exponential(rate, skewweight)

    See Goerg 2011 and also Kaarik et al (2023) - https://arxiv.org/pdf/2307.05644.pdf

    Args:
        rate (float or Tensor): rate of log of the distribution
        skewweight (float or Tensor): skewweight ("gamma") of the Lambert W x Normal distribution.
          If 0., then it reduces to a Normal(loc, scale) distribution.
    """
    arg_constraints = {
        "rate": td.constraints.nonnegative,
        "skewweight": td.constraints.real,
    }
    support = td.constraints.nonnegative
    has_rsample = True

    def __init__(self, rate, skewweight, validate_args=None):
        self.skewweight = skewweight
        self.rate = rate
        super().__init__(
            base_distribution=td.Exponential(rate=rate, validate_args=validate_args),
            transforms=transforms.LambertWSkewTransform(
                shift=0.0, scale=1.0 / rate, skewweight=skewweight
            ),
            validate_args=validate_args,
        )

    def expand(self, batch_shape, _instance=None):
        new = self._get_checked_instance(LambertWNormal, _instance)
        return super().expand(batch_shape, _instance=new)

    @property
    def mean(self):
        # See p13 in https://arxiv.org/pdf/2307.05644.pdf
        mean_y = 1.0 / (self.rate * (1.0 - self.skewweight) ** 2)
        return torch.where(mean_y > 0, mean_y, torch.nan)

    @property
    def mode(self):
        return torch.zeros_like(self.rate)

    @property
    def variance(self):
        # See p13 in https://arxiv.org/pdf/2307.05644.pdf
        # variance = EY^2 - (mean)**2
        denom = 1.0 - 2 * self.skewweight
        second_moment = 2.0 / (self.rate.pow(2.0) * torch.pow(denom, 2.0 + 1.0))
        variance = second_moment - self.mean ** 2
        return variance
